iceoryx_utils/test/moduletests/test_resizeable_lockfree_queue.cpp

```diff
@@ -0,0 +1,402 @@
+}
+
+TYPED_TEST(ResizeableLockFreeQueueTest, setCapacityToMaxCapacity)
+{
+    typename TestFixture::Queue q(0);
+    constexpr auto MAX_CAP = TestFixture::Queue::MAX_CAPACITY;
+    EXPECT_TRUE(q.setCapacity(MAX_CAP));
+    EXPECT_EQ(q.capacity(), MAX_CAP);
+}
+
+TYPED_TEST(ResizeableLockFreeQueueTest, setCapacityToHalfOfMaxCapacityAndFillIt)
+{
+    auto& q = this->queue;
+    constexpr auto MAX_CAP = TestFixture::Queue::MAX_CAPACITY;
+    uint64_t newCap = MAX_CAP / 2;
+    EXPECT_EQ(q.setCapacity(newCap), true);
+    EXPECT_EQ(q.capacity(), newCap);
+
+    uint64_t element = 0;
+    while (q.tryPush(element++))
+        ;
+
+    EXPECT_EQ(q.capacity(), newCap);
+    EXPECT_EQ(q.size(), newCap);
+    EXPECT_EQ(element, newCap + 1);
+}
+
+TYPED_TEST(ResizeableLockFreeQueueTest, setCapacityFromHalfOfMaxCapacityToMaxCapacity)
+{
+    auto& q = this->queue;
+    constexpr auto MAX_CAP = TestFixture::Queue::MAX_CAPACITY;
+    uint64_t cap = MAX_CAP / 2;
+    EXPECT_EQ(q.setCapacity(cap), true);
+    EXPECT_EQ(q.capacity(), cap);
+
+    uint64_t element = 0;
+    while (q.tryPush(element++))
+        ;
+
+    EXPECT_EQ(q.capacity(), cap);
+    EXPECT_EQ(q.size(), cap);
+    EXPECT_EQ(element, cap + 1);
+
+    EXPECT_EQ(q.setCapacity(MAX_CAP), true);
+    EXPECT_EQ(q.capacity(), MAX_CAP);
+    EXPECT_EQ(q.size(), cap);
+
+    --element; // compensate the last failed tryPush
+    while (q.tryPush(element++))
+        ;
+
+    // we want to find all elements we pushed
+    for (element = 0; element < MAX_CAP; ++element)
+    {
+        auto result = q.pop();
+        ASSERT_TRUE(result.has_value());
+        EXPECT_EQ(result.value(), element);
+    }
+}
+
+TYPED_TEST(ResizeableLockFreeQueueTest, setCapacityOfFullQueueToHalfOfMaxCapacity)
+{
+    auto& q = this->queue;
+    constexpr auto MAX_CAP = TestFixture::Queue::MAX_CAPACITY;
+    uint64_t cap = MAX_CAP / 2;
+
+    uint64_t element = 0;
+    while (q.tryPush(element++))
+        ;
+    EXPECT_EQ(q.capacity(), MAX_CAP);
+    EXPECT_EQ(q.size(), MAX_CAP);
+
+    EXPECT_TRUE(q.setCapacity(cap));
+    EXPECT_EQ(q.capacity(), cap);
+    EXPECT_EQ(q.size(), cap);
+
+    if (cap == 0)
+    {
+        return;
+    }
+    // the least recent values are removed due to the capacity being decreased
+    for (element = cap; element < MAX_CAP; ++element)
+    {
+        auto result = q.pop();
+        ASSERT_TRUE(result.has_value());
+        EXPECT_EQ(result.value(), element);
+    }
+}
+
+// note this is one of the most general cases and necessary to test:
+// decreasing the capacity starting with a partially filled queue and checking whether the last values
+// remain (and the others are removed)
+TYPED_TEST(ResizeableLockFreeQueueTest, DecreaseCapacityOfAPartiallyFilledQueue)
+{
+    auto& q = this->queue;
+    constexpr auto MAX_CAP = TestFixture::Queue::MAX_CAPACITY;
+    iox::cxx::vector<int, MAX_CAP> removedElements;
+
+    uint64_t cap = MAX_CAP / 2;
+
+    EXPECT_TRUE(q.setCapacity(cap));
```

Comment:
You already tested this various times in the previous tests. Therefore I would get rid of all EXPECT_* which verifies already tested behavior. You can assume that everything which have been tested and verified before is working correctly.

This would help a lot when trying to understand the test since you see where the setup of the test is, where the actual action is happening and how you verify the correctness of the behavior. Here it seems a little bit like you  test everything.

Additionally, all our tests should be written according to the arrange, act, assert pattern: https://medium.com/@pjbgf/title-testing-code-ocd-and-the-aaa-pattern-df453975ab80

Could you please adjust the tests here.

Reply:
They are arrange act assert, but there are more complex expectations in the assert. Create a suitably filled queue (Arrange), call setCapacity (Act), check whether the result satisfies the expectations (assert). The last part is a bit complicated, mainly due to the checks for removed elements but I may be able to distinguish the phases a bit better.

I also have to recheck some seemingly tested behavior (but an assert may be needed), because some subsequent checks have no hope succeeding in some cases if a previous one failed.

Reply:
Reduced the redundancy in checks during the test setups (arrange phase).